File: factors.c
// factors.c
// A program that factorizes natural numbers into a product of two smaller numbers

#include <stdio.h>
#include <stdlib.h>

// A function that prints a factorization of a number n
void print_factorization(unsigned long long n) {
  // Find the smallest divisor of n
  unsigned long long divisor = 2;
  while (divisor * divisor <= n) {
    if (n % divisor == 0) {
      // Print the factorization
      printf("%llu=%llu*%llu\n", n, divisor, n / divisor);
      return;
    }
    divisor++;
  }
  // If n is prime, print n=1*n
  printf("%llu=1*%llu\n", n, n);
}

// A function that reads a file and factorizes each number in it
void factorize_file(char *filename) {
  // Open the file
  FILE *file = fopen(filename, "r");
  if (file == NULL) {
    printf("Error: cannot open file %s\n", filename);
    exit(1);
  }
  // Read each number and factorize it
  unsigned long long number;
  while (fscanf(file, "%llu", &number) == 1) {
    print_factorization(number);
  }
  // Close the file
  fclose(file);
}

// The main function that takes a file name as an argument
int main(int argc, char *argv[]) {
  // Check if the argument is given
  if (argc != 2) {
    printf("Usage: factors <file>\n");
    exit(1);
  }
  // Factorize the file
  factorize_file(argv[1]);
  return 0;
}

File: factors.py
#!/usr/bin/python3

'''# factors.py
# A program that can factorize numbers from a file
# Usage: factors <file>
# where <file> is a file containing natural numbers to factor
# One number per line
# Output format: n=p*q
# one factorization per line
# p and q don’t have to be prime numbers
'''
import sys

# Check if a file name is given as an argument
if len(sys.argv) != 2:
    print("Usage: factors <file>")
    sys.exit(1)

# Open the file for reading
try:
    file = open(sys.argv[1], "r")
except IOError:
    print("Error: cannot open file", sys.argv[1])
    sys.exit(2)

# Read the file line by line
for line in file:
    # Strip the newline character and convert to integer
    n = int(line.strip())
    # Find a factor of n by trying from 2 to sqrt(n)
    for p in range(2, int(n**0.5) + 1):
        # If p divides n, then q = n / p is another factor
        if n % p == 0:
            q = n // p
            # Print the factorization
            print(f"{n}={p}*{q}")
            # Break the loop
            break
    else:
        # If no factor is found, n is prime
        print(f"{n} is prime")

# Close the file
file.close()
